### PR TITLE
front: relabel duplicate trigrams in nge node import

### LINKED ISSUE #13276: Handle different nodes sharing the same trigram when importing NGE dtos
### Description and goal

Part of [#osrd-confidential/1162: Improve NGE import for RNE](https://github.com/osrd-project/osrd-confidential/issues/1162)

Some nodes in NGE dtos from RNE share the same trigram, but have different ids. They should be handled as different nodes when importing such dtos. To this end, we need to modify their trigrams, as we can only have node per unique trigram : "PLY" would for example become "PLY-1", "PLY-2", "PLY-3", ...

### Acceptance criteria

- Nodes with the same trigrams but different ids are created as different nodes with unique trigrams

### PR DESCRIPTION
**Pull request by Synar**
Close [#13276](https://github.com/OpenRailAssociation/osrd/issues/13276)

Todo before merging : add a test case

### CODE REVIEW

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    { ...baseNode, trigram: 'DDD', ngeId: 2 },
+    { ...baseNode, trigram: 'BBB', ngeId: 3 },
+    { ...baseNode, trigram: 'AAA', ngeId: 4 },
+    { ...baseNode, trigram: 'AAA', ngeId: 5 },
+    { ...baseNode, trigram: 'CCC', ngeId: 6 },
+    { ...baseNode, trigram: 'DDD', ngeId: 7 },
+  ];
+  const nodesWUniqueTrigrams = relabelDuplicateTrigrams(nodes);
+
+  it('shoud not add or remove nodes', () => {
```
Comment (louisgreiner) [inline front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]:
```suggestion
  it('should not add or remove nodes', () => {
```

nit: typo

Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: Thanks!

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  const nodes = [
+    { ...baseNode, betriebspunktName: 'AAA', id: 1 },
+    { ...baseNode, betriebspunktName: 'DDD', id: 2 },
+    { ...baseNode, betriebspunktName: 'BBB', id: 3 },
+    { ...baseNode, betriebspunktName: 'AAA', id: 4 },
+    { ...baseNode, betriebspunktName: 'AAA', id: 5 },
+    { ...baseNode, betriebspunktName: 'CCC', id: 6 },
+    { ...baseNode, betriebspunktName: 'DDD', id: 7 },
+  ];
+  const nodesWUniqueTrigrams = relabelDuplicateTrigrams(nodes);
```
Comment (emersion) [inline front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: Instead of a unit test, I would prefer a new test in `src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeToOsrd.spec.ts`. These test the whole import machinery and are easier to maintain.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts
Comment (emersion) [inline front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts:1]: Not a fan of adding a new `ngeImportUtils.ts` file: this is a pretty small function which could be in `ngeToOsrd.ts`.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts [hunk 3]
```diff
@@ -946,8 +946,9 @@ export const handleOperation = async ({
 };
 
 export const convertNgeDtoToOsrd = (dto: NetzgrafikDto) => {
-  const macroNodes: MacroNodeForm[] = [];
-  for (const node of dto.nodes) {
+  const nodesWUniqueTrigrams = relabelDuplicateTrigrams(dto.nodes);
+  const macroNodes: Omit<NodeIndexed, 'dbId'>[] = [];
```
Comment (emersion) [inline front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts]: Can we avoid using `NodeIndexed`? This type is specific to `MacroEditorState`, which isn't involved here.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import type { NodeDto } from '../NGE/types';
+
+export const relabelDuplicateTrigrams = (nodes: NodeDto[]): NodeDto[] => {
+  const trigramsToIds: Record<string, number[]> = {};
+  for (const node of nodes) {
+    if (!trigramsToIds[node.betriebspunktName]) trigramsToIds[node.betriebspunktName] = [];
+    trigramsToIds[node.betriebspunktName].push(node.id);
+  }
+
+  const nodesWUniqueTrigrams = [];
```
Comment (emersion) [inline front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts]: Can we spell out "with" instead of using a single letter? Or just name it `dedupNodes`?

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import type { NodeDto } from '../NGE/types';
+
+export const relabelDuplicateTrigrams = (nodes: NodeDto[]): NodeDto[] => {
+  const trigramsToIds: Record<string, number[]> = {};
```
Comment (emersion) [inline front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts]: Can we use a `Map` instead? `Record` has a number of issues with user-provided keys.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  for (const node of nodes) {
+    if (!trigramsToIds[node.betriebspunktName]) trigramsToIds[node.betriebspunktName] = [];
+    trigramsToIds[node.betriebspunktName].push(node.id);
+  }
+
+  const nodesWUniqueTrigrams = [];
+  for (const node of nodes) {
```
Comment (emersion) [inline front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts]: This could be a `.map()`.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts
Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts:1]: ngeToOsrd is very long and cumbersome though. I would really like to break it down in smaller pieces.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts (hunk 2 above)
Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: I would prefer both tbh. I do intend on writing a test case with duplicated nodes, but tests with large json files are imho harder to understand and debug.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts (hunk 4 above)
Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts]: I'm fine with dedupNodes ^^

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts (hunk 5 above)
Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts]: Sure!

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts (hunk 6 above)
Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts]: Sure. I like the symmetry of the 2 loops being the same, but I'm fine with using a map too.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts (hunk 3 above)
Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts]: Sure! I used to need it for a previous version of the code, and I just forgot to revert it to MacroNodeForm.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts
Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts:1]: I did assume I would have to add more functions in that file for the other import improvements, but none of them ended up requiring it, so this may not be the right way to break down the file. Still I believe we need to refactor it in some way. But we can talk about this in a future pr.

Comment (emersion) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeImportUtils.ts:1]: I agree that `ngeToOsrd.ts` is a tad long and could be split up, but I don't think moving lower-level functions to a separate file is a good way to split up a file in general. I'd prefer to split up per kind of object for instance: `ngeToOsrd/node.ts`, `ngeToOsrd/trainrun.ts`, and so on. But yeah, that's orthogonal to this PR and should be discussed elsewhere.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts (hunk 2 above)
Comment (emersion) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: The function is not especially complicated nor has a huge number of edge cases, so I don't believe a unit test is warranted here.

Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: It's not extremely complex, but it's not a one liner either. I think it does enough things to warrant a simple test, and it's not like our unit tests are particularly slow. I don't believe unit tests add a ton of complexity to the codebase either : on the contrary, I believe having unit tests help anyone who wants to modify the code later clearly understand what it was trying to achieve.

Comment (emersion) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: Unit tests have a maintenance cost: they make refactorings more difficult because more code depends on internals.

Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: If someone wants to refactor away the function, getting rid of the unit test is extremely straight forward though. And reading the unit test helps them get exactly what the function was achieving at a glance. I think it's an overall net positive.

Comment (emersion) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: Well, usually a refactor will change the inputs/outputs or scope, and we'd need to adapt the tests, not get rid of them. I've done that countless times on OSRD and it's pretty time-consuming.

See also: https://grugbrain.dev/#grug-on-testing

Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: I like grugbrain too, but I would still rather have this test that not =p I really don't believe a lot of time would be spent refactoring this code any time soon.

Proper duplicated node handling is also an edge case that would be easy to accidentally get rid of when refactoring the json test files, since they're quite unwieldy to read and write.

Comment (emersion) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: Regardless of whether this specific instance will in the future end up requiring some work when refactoring, my goal here is to establish best practices. I don't really care whether this specific instance gets merged, but I do care about such overly low-level unit tests not growing in numbers in future PRs.

Finding good boundaries for tests is important to me.

> Proper duplicated node handling is also an edge case that would be easy to accidentally get rid of when refactoring the json test files

Each test can easily describe what it's testing for. Test files are in a standard NGE or editoast format.

> since they're quite unwieldy to read and write

JSON files can be opened directly straight in NGE for visualization. Generated editoast payloads are readable and familiar.

Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: > Regardless of whether this specific instance will in the future end up requiring some work when refactoring, my goal here is to establish best practices. I don't really care whether this specific instance gets merged, but I do care about such overly low-level unit tests not growing in numbers in future PRs.

I guess we can talk next Tuesday about how many unit tests we want


> Each test can easily describe what it's testing for. Test files are in a standard NGE or editoast format.

If we write a proper test for each case, we would have to write a lot of very long jsons. Here I went from 7 nodes in the unit test to 3 nodes in the json file, and there are still 500 lines of json. 

The fact they can be visualized doesn't make them easier to read, write or refactor either. Refactoring the previous json into the one way and round trip jsons was a huge pain, and it took me quite a few iterations to get it right. Probably longer than writing this entire pr.

> Generated editoast payloads are readable and familiar.

Maybe to you =p (But the output files are indeed more concise than the input ones and contain less indirection, which helps.)

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts [hunk 7]
```diff
@@ -945,9 +945,31 @@ export const handleOperation = async ({
...
+  const trigramsToIds = new Map<string, number[]>();
+  for (const node of nodes) {
+    if (!trigramsToIds.has(node.betriebspunktName)) trigramsToIds.set(node.betriebspunktName, []);
+    trigramsToIds.get(node.betriebspunktName)!.push(node.id);
+  }
+
+  return nodes.map((node) => {
+    const trigramIds = trigramsToIds.get(node.betriebspunktName)!;
+    if (trigramIds.length == 1) return node;
+    else {
```
Comment (emersion) [inline front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts]: Nit: this `else` is unnecessary, since there is a `return` above.

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts [hunk 8]
```diff
@@ -945,9 +945,31 @@ export const handleOperation = async ({
...
+      const idIndex = trigramIds.findIndex((id) => id === node.id);
```
Comment (emersion) [inline front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts]: Nit: would be slightly more human-readable to add `+ 1` here, to have these nodes start at `XXX-1` instead of `XXX-0`.

**Review by emersion — APPROVED** [review]
LGTM apart from the remaining opened discussions!

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts (hunk 7 above)
Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts]: Oh indeed thanks, I refactored from for to .map too fast ^^

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts (hunk 8 above)
Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts]: Sure!

File: front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts (hunk 2 above)
Comment (Synar) [reply front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/__tests__/ngeImportUtils.spec.ts]: Well, let this code live here if nowhere else, for anyone that shall wonder what the hell we were bikeshedding about :
```
describe('relabelDuplicateTrigrams', () => {
  const baseNode = {
    fullName: '',
    positionX: 10,
    positionY: 20,
    ports: [],
    transitions: [],
    connections: [],
    resourceId: 42,
    perronkanten: 1,
    connectionTime: 2,
    trainrunCategoryHaltezeiten: {},
    symmetryAxis: 1,
    warnings: [],
    labelIds: [8],
  };
  const nodes = [
    { ...baseNode, betriebspunktName: 'AAA', id: 1 },
    { ...baseNode, betriebspunktName: 'DDD', id: 2 },
    { ...baseNode, betriebspunktName: 'BBB', id: 3 },
    { ...baseNode, betriebspunktName: 'AAA', id: 4 },
    { ...baseNode, betriebspunktName: 'AAA', id: 5 },
    { ...baseNode, betriebspunktName: 'CCC', id: 6 },
    { ...baseNode, betriebspunktName: 'DDD', id: 7 },
  ];
  const nodesWUniqueTrigrams = relabelDuplicateTrigrams(nodes);

  it('should not add or remove nodes', () => {
    expect(nodesWUniqueTrigrams.length).toBe(7);
  });

  it('should ignore unique trigrams', () => {
    expect(nodesWUniqueTrigrams[2]).toEqual({ ...baseNode, betriebspunktName: 'BBB', id: 3 });
    expect(nodesWUniqueTrigrams[5]).toEqual({ ...baseNode, betriebspunktName: 'CCC', id: 6 });
  });

  it('should relabel duplicated trigrams', () => {
    expect(nodesWUniqueTrigrams[0]).toEqual({ ...baseNode, betriebspunktName: 'AAA-0', id: 1 });
    expect(nodesWUniqueTrigrams[1]).toEqual({ ...baseNode, betriebspunktName: 'DDD-0', id: 2 });
    expect(nodesWUniqueTrigrams[3]).toEqual({ ...baseNode, betriebspunktName: 'AAA-1', id: 4 });
    expect(nodesWUniqueTrigrams[4]).toEqual({ ...baseNode, betriebspunktName: 'AAA-2', id: 5 });
    expect(nodesWUniqueTrigrams[6]).toEqual({ ...baseNode, betriebspunktName: 'DDD-1', id: 7 });
  });
});
```
